geo/src/algorithm/stitch.rs

```diff
@@ -0,0 +1,484 @@
+use std::collections::HashMap;
+
+use geo_types::{Coord, Line, LineString, MultiPolygon, Polygon, Triangle};
+
+use crate::{Contains, CoordsIter, GeoFloat, LinesIter, Winding};
+
+// ========= Error Type ============
+
+#[derive(Debug)]
+pub enum LineStitchingError {
+    IncompleteRing,
+    InvalidGeometry,
+    MissingParent,
+    NoExtremum,
+}
+
+impl std::fmt::Display for LineStitchingError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{self:?}")
+    }
+}
+
+impl std::error::Error for LineStitchingError {}
+
+pub(crate) type PolygonStitchingResult<T> = Result<T, LineStitchingError>;
+
+// ========= Main Algo ============
+
+pub trait Stitch<T: GeoFloat> {
+    fn stitch_together(&self) -> PolygonStitchingResult<MultiPolygon<T>>;
+}
+
+macro_rules! impl_stitch {
+    ($type:ty => $self:ident
+     fn stitch_together(&self) -> PolygonStitchingResult<MultiPolygon<T>> {
+         $($impls:tt)*
+     }) => {
+       impl<T> Stitch<T> for &[&$type]
+       where
+           T: GeoFloat,
+       {
+           fn stitch_together(&$self) -> PolygonStitchingResult<MultiPolygon<T>> {
+                $($impls)*
+           }
+       }
+
+       impl<T> Stitch<T> for Vec<&$type>
+       where
+           T: GeoFloat,
+       {
+           fn stitch_together(&$self) -> PolygonStitchingResult<MultiPolygon<T>> {
+               $self.as_slice().stitch_together()
+           }
+       }
+
+       impl<T> Stitch<T> for &[$type]
+       where
+           T: GeoFloat,
+       {
+           fn stitch_together(&$self) -> PolygonStitchingResult<MultiPolygon<T>> {
+               $self.iter().collect::<Vec<_>>().stitch_together()
+           }
+       }
+
+       impl<T> Stitch<T> for Vec<$type>
+       where
+           T: GeoFloat,
+       {
+           fn stitch_together(&$self) -> PolygonStitchingResult<MultiPolygon<T>> {
+               $self.iter().collect::<Vec<_>>().stitch_together()
+           }
+       }
+     };
+}
+
+impl_stitch! {
+    Polygon<T> => self
+    fn stitch_together(&self) -> PolygonStitchingResult<MultiPolygon<T>> {
+        // 🚧🚧 TODO 🚧🚧 : it might make sense to have a separate method in this trait which
+        // doesn't do the fixup and instead assumes that the polygon is well formed to prevent
+        // extensive cloning
+        let polys = self
+            .iter()
+            .map(|&poly| fix_orientation(poly.clone()))
+            .collect::<Vec<_>>();
```

Comment:
We could also just do a conditional clone here which should improve the situation already quiet a bit

Reply:
Actually, `make_ccw` which is used for this only has an effect if the winding order is `cw`. Checking this again beforehand would just double the costs for no improvements